api/eds.proto

```diff
@@ -0,0 +1,188 @@
+  // server. The steps are:
+  // 1, The management server is configured with per cluster/zone/load metric
+  //    capacity configuration. The capacity configuration definition is
+  //    outside of the scope of this document.
+  // 2. Envoy issues LoadBalanceRequest for a list of clusters it wants to load
+  //    balance to (instead of using a basic EndpointDiscoveryRequest).
+  // 3. Once a connection establishes, the management server publishes
+  //    LoadBalanceResponse for all clusters requested in the first
+  //    LoadBalanceRequest. This message contains per cluster/per Locality
+  //    information.
+  // 4. For each cluster, Envoy load balances incoming traffic to upstream hosts
+  //    based on per-zone weights and/or per-instance weights (if specified)
+  //    based on intra-zone LbPolicy.
+  // 5. When upstream hosts reply, they optionally add header <define header
+  //    name> with ASCII representation of EndpointLoadMetricStats.
+  // 6. Envoy aggregates load reports over the period of time given to it in
+  //    LoadBalanceResponse.load_reporting_interval. This includes aggregation
+  //    stats Envoy maintains by itself (total_requests, rpc_errors etc.) as
+  //    well as load metrics from upstream hosts.
+  // 7. When the timer of load_reporting_interval expires, Envoy sends new
+  //    LoadBalanceRequest filled with load reports for each cluster.
+  // 8. The management server uses the load reports from all reported Envoys
+  //    from around the world, computes global assignment and prepares traffic
+  //    assignment destined for each zone Envoys are located in. Goto 3.
+  // TODO(htuch): Add @amb67's diagram.
+  rpc BalanceLoad(stream LoadBalanceRequest)
+      returns (stream LoadBalanceResponse) {
+  }
+}
+
+message EndpointDiscoveryRequest {
+  repeated string cluster_name = 1;
+  Node node = 2;
+}
+
+message LbEndpoint {
+  // Endpoint host address.
+  Address address = 1;
+
+  // Optional health status when known and supplied by EDS server.
+  HealthStatus health_status = 2;
+
+  // The optional canary status of the upstream host. Envoy uses the canary
+  // status for various statistics and load balancing tasks documented
+  // elsewhere.
+  google.protobuf.BoolValue canary = 3;
+  // The optional load balancing weight of the upstream host, in the range 1 -
+  // 100. Envoy uses the load balancing weight in some of the built in load
+  // balancers. All load balancing weights in a locality must sum to 100%.
+  // The effective load balancing weight is load_balancing_weight multiplied by
+  // the locality wide load_balancing_weight. If unspecified, each host is
+  // presumed to have equal weight in a locality.
+  google.protobuf.UInt32Value load_balancing_weight = 4;
+}
+
+// All endpoints belonging to a Locality.
+message LocalityEndpoints {
+  Locality locality = 1;
+  repeated LbEndpoint lb_endpoints = 2;
+
+  // Optional: Per region/zone/sub_zone weight - range 1-100. All load balancing
```

Comment:
In the current code, weight does not need to sum to 100 for each host in the locality. It works like most traditional HW LBs where % of traffic is equal to (node weight) / (total of all weights). The current code also has a pretty sub-optimal implementation of weighted least request which we only use for red line testing. I'm fine with changing the semantics but I would call it out as a diff.